util/generate_haddock3_catalog.py

```diff
@@ -47,30 +48,103 @@
 import logging
 from math import isnan
 from pathlib import Path
+import re
 import sys
 from yaml import dump, load, Loader
 
 from haddock.modules import modules_category
-from haddock import config_expert_levels
+from haddock import config_expert_levels, _hidden_level
 
 def argparser_builder():
     parser = argparse.ArgumentParser()
     parser.add_argument('--out_dir', type=Path, default='public/catalog')
     parser.add_argument('--root_url', type=Path, default='/catalog')
     return parser
 
+def collapse_expandable(config):
+    """
+    The haddock3 defaults.yaml files define complex shape inside the parameter name.
+    The workflow builder uses type:array, type:object from JSON schema to define complex data types.
+
+    The method converts the shape in parameter name into a shortened parameter name and a shape
+    1. arrays of scalars (X_1 -> X:[1]),
+    2. arrays of objects (X_Y_1 -> X:[{Y}]),
+    3. array of arrays of scalars (X_1_1 -> X:[[1]]) and
+    4. array of arrays of object (X_Y_1_1 -> X:[[{Y}]])
+
+    """
+    array_of_scalar = r'(\w+)_1'
+    array_of_object = r'(\w+)_(\w+)_1'
+    array_of_array_of_scalar = r'(\w+)_1_1'
+    array_of_array_of_object = r'(\w+)_(\w+)_1_1'
+    skip = r'\w+?(_\d\d?)(_\d\d?)?'
+    must_be_array_of_scalar = {'mol_fix_origin_1', 'mol_shape_1'}
+    new_config = {}
+    for k, v in config.items():
+        logging.info(f'Processing var: {k}')
+        if (match := re.match(skip, k)) and not (match.groups()[0] == '_1' and (match.groups()[1] in {None,'_1'})):
+            logging.info(f'SKipping {k} as their schema will be captured by first index.')
+            continue
+            # TODO nice to store maxitems with type:array
+            # for example rair_end_20_1 -> outer array should have maxitems:20
+            # for example int_20_20 ->  both array should have maxitems:20
+        elif match := re.match(array_of_array_of_object, k):
+            p, n = match.groups()
+            if p not in new_config:
+                new_config[p] = {'dim': 2, 'properties': {}, 'type': 'list'}
+            new_config[p]['properties'][n] = v
+            if 'group' in v:
+                # Move group from nested prop to outer array
+                new_config[p]['group'] = v['group']
+                del v['group']
```

Comment:
So as a side-effect of calling this function, the incoming config is modified?

Reply:
The array will be in that group and the scalar value is a child of that array. So the scalar is already in the group, by keeping a group on the scalar it would be rendered in its own collapsible subform.

You would then have a collapsible subforms inside an array itemwhich was already inside a collapsible subform felt a bit to much folding going on.

Also the [grouper](https://github.com/i-VRESSE/workflow-builder/blob/generate-catalog-with-tomlschema/src/grouper.ts) currently only works on parameters that are not nested.